src/Rector/Architecture/RepositoryAsService/ServiceLocatorToDIRector.php

```diff
@@ -0,0 +1,129 @@
+        PropertyFetchNodeFactory $propertyFetchNodeFactory,
+        RepositoryForDoctrineEntityProviderInterface $repositoryForDoctrineEntityProvider,
+        ClassPropertyCollector $classPropertyCollector
+    ) {
+        $this->methodCallAnalyzer = $methodCallAnalyzer;
+        $this->propertyFetchNodeFactory = $propertyFetchNodeFactory;
+        $this->repositoryForDoctrineEntityProvider = $repositoryForDoctrineEntityProvider;
+        $this->classPropertyCollector = $classPropertyCollector;
+    }
+
+    public function isCandidate(Node $node): bool
+    {
+        if (! $this->methodCallAnalyzer->isMethod($node, 'getRepository')) {
+            return false;
+        }
+
+        $className = $node->getAttribute(Attribute::CLASS_NAME);
+
+        if ($className === null) {
+            return false;
+        }
+
+        return ! Strings::endsWith($node->getAttribute(Attribute::CLASS_NAME), 'Repository');
+    }
+
+    public function refactor(Node $node): ?Node
+    {
+        $this->classPropertyCollector->addPropertyForClass(
+            (string) $node->getAttribute(Attribute::CLASS_NAME),
+            [$this->repositoryFqn($node)],
+            $this->repositoryVariableName($node)
+        );
+
+        return $this->propertyFetchNodeFactory->createLocalWithPropertyName(
+            $this->repositoryVariableName($node)
+        );
+    }
+
+    private function repositoryVariableName(Node $node): string
+    {
+        return lcfirst($this->repositoryShortName($node));
+    }
+
+    private function repositoryShortName(Node $node): string
+    {
+        $nameSpaceParts = explode('\\', $this->repositoryFqn($node));
+        return end($nameSpaceParts);
+    }
+
+    private function repositoryFqn(Node $node): string
+    {
+        $entityFqnOrAlias = $this->entityFqnOrAlias($node);
+
+        $repositoryClassName = $this->repositoryForDoctrineEntityProvider->provideRepositoryForEntity(
+            $entityFqnOrAlias
+        );
+
+        if ($repositoryClassName === null) {
+            throw new RectorProviderException(sprintf(
+                'A repository was not provided for "%s" entity by your "%s" class.',
+                $entityFqnOrAlias,
+                get_class($this->repositoryForDoctrineEntityProvider)
+            ));
+        }
+
+        return $repositoryClassName;
+    }
+
+    /**
+     * @throws ShouldNotHappenException
+     */
+    private function entityFqnOrAlias(Node $node): string
+    {
+        $repositoryArgument = $node->args[0]->value;
+
+        if ($repositoryArgument instanceof String_) {
+            return $repositoryArgument->value;
+        }
+
+        if ($repositoryArgument instanceof ClassConstFetch && $repositoryArgument->class instanceof Name) {
+            return $repositoryArgument->class->getAttribute(Attribute::RESOLVED_NAME)->toString();
+        }
+
+        throw new ShouldNotHappenException('Unable to resolve repository argument');
```

Comment:
This would be better to test, since there are many paths to take and end up with.

Reply:
I don't understand what do you want me to do, add test with wrong repository parameter to test it throws this exception?

Reply:
I meant to test all these private methods for all  paths

Reply:
Should I separate this one too?